samza-sql-avro/src/main/java/org/apache/samza/sql/avro/AvroRelToSamzaMsgConverterFactory.java

```diff
@@ -0,0 +1,124 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.avro;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.reflect.ReflectData;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.sql.interfaces.RelToSamzaMsgConverter;
+import org.apache.samza.sql.interfaces.RelToSamzaMsgConverterFactory;
+
+
+public class AvroRelToSamzaMsgConverterFactory implements RelToSamzaMsgConverterFactory {
+
+  private Map<String, Schema> schemas = new HashMap<>();
+
+  @Override
+  public RelToSamzaMsgConverter create(Properties config) {
+    return new AvroRelToSamzaMsgConverter(config);
+  }
+
+  private class AvroRelToSamzaMsgConverter implements RelToSamzaMsgConverter {
+
+    private final Schema arraySchema;
+    private final Schema mapSchema;
+    private Map<String, Schema> streamToSchemaCache = new HashMap<>();
+
+    public class TestClass {
```

Comment:
It looks like you're using this to programmatically derive the avro type? Is there a more direct way to achieve this (I'm not too familiar with Avro)? In any case, could we make this private and then set it as accessible via reflection if required? The benefit is that we don't expose this externally to the class.

Reply:
Agreed! Since the avro schema corresponding to the collection and Map is going to be static. I am leaning towards making the avro schema corresponding to these as constants. Thoughts?